### PR TITLE
(maint) Enable some rubocop checks for 5.5.x

### PR DESCRIPTION
**Pull request by joshcooper**
Enables some useful rubocops. Some notable ones are:

* ~`Performance/Caller`~ dropped because ruby 1.9.3 doesn't have a `Kernel.caller(Range)` method
* `Performance/Casecmp`
* `Performance/CompareWithBlock`
* `Lint/DuplicateMethods`

The full catalog benchmark is about the same for MRI 2.3.8 and JRuby 9.1.17.0 before and after this PR:

## MRI 2.3.8

```$ bx rake benchmark:full_catalog:run ITERATIONS=10```

Before:

```
> avg:       1.712000   0.048000   1.760000 (  1.761256)
```

After:

```
> avg:       1.734000   0.049000   1.783000 (  1.783701)
```

## JRuby 9.1.17.0

```$ bx rake benchmark:full_catalog:run ITERATIONS=100```

Before:

```
> avg:       3.207600   0.071700   3.279300 (  2.114744)
```

After:

```
> avg:       3.160700   0.072800   3.233500 (  2.101808)
```

### CODE REVIEW

**Comment by GabrielNagy** [comment]
jenkins please test this on aix61-POWERfa aix71-POWERfa aix72-POWERfa amazon6-64a amazon7-64a arista4-32a debian8-64a debian8-32a debian9-64a debian9-32a fedora28-64a fedora29-64a fedora30-64a osx1012-64a osx1013-64a osx1014-64a redhat5-64a centos6-64a centos6-32a redhat7-64a redhat7-POWERa redhat7-AARCH64a redhat8-64a redhatfips7-64a sles11-64a sles11-32a sles12-64a sles12-POWERa sles15-64a solaris10-64a solaris11-64a solaris114-64a ubuntu1404-64a ubuntu1404-32a ubuntu1604-64a ubuntu1604-32a ubuntu1604-POWERa ubuntu1804-64a windows10ent-64a windows10ent-32a windows2008r2-64a windows2012r2-64a windows2016-64a windows2019-64a

File: lib/puppet/type.rb [hunk 1]
```diff
@@ -1909,9 +1909,7 @@ def self.doc
         # Since we're mixing @doc with text from other sources, we must normalize
         # its indentation with scrub. But we don't need to manually scrub the
         # provider's doc string, since markdown_definitionlist sanitizes its inputs.
-        scrub(@doc) + "Available providers are:\n\n" + parenttype.providers.sort { |a,b|
-          a.to_s <=> b.to_s
-        }.collect { |i|
+        scrub(@doc) + "Available providers are:\n\n" + parenttype.providers.sort_by(&:to_s).collect { |i|
```
Comment (gimmyxd) [inline lib/puppet/type.rb]: This should be wrapped in a `do..end` instead of `{..}` because the block spwans on multiple lines

File: lib/puppet/util.rb [hunk 2]
```diff
@@ -323,7 +323,7 @@ def path_to_uri(path)
     params = { :scheme => 'file' }
 
     if Puppet.features.microsoft_windows?
-      path = path.gsub(/\\/, '/')
+      path = path.tr('\\', '/')
```
Comment (gimmyxd) [inline lib/puppet/util.rb]: You could use `path.tr!`

File: lib/puppet/face/module/upgrade.rb [hunk 3]
```diff
@@ -67,7 +67,7 @@
     end
 
     when_invoked do |name, options|
-      name = name.gsub('/', '-')
+      name = name.tr('/', '-')
```
Comment (gimmyxd) [inline lib/puppet/face/module/upgrade.rb]: You could use `name.tr!`

File: lib/puppet/indirector/hiera.rb [hunk 4]
```diff
@@ -69,6 +69,8 @@ def convert_merge(merge)
     end
   end
 
+  public
```
[the remark this answers is not in the document]
Comment (gimmyxd) [reply lib/puppet/indirector/hiera.rb]: Also i'm leaning towards having class methods in a `class << self` at top because it will help us keep the class methods grouped.

File: lib/puppet/provider/augeas/augeas.rb [hunk 5]
```diff
@@ -103,7 +103,7 @@ def parse_commands(data)
             inSingleTick = !inSingleTick if ch == "'"
             inDoubleTick = !inDoubleTick if ch == "\""
             fail(_("unmatched [")) if nbracket < 0
-          end until ((nbracket == 0 && !inSingleTick && !inDoubleTick && (ch =~ /\s/)) || sc.eos?)
+          end until ((nbracket == 0 && !inSingleTick && !inDoubleTick && (ch =~ /\s/)) || sc.eos?) # rubocop:disable Lint/Loop
```
Comment (gimmyxd) [inline lib/puppet/provider/augeas/augeas.rb]: Should we change this to` loop do end` since is similar to `begin end until`?

File: lib/puppet/util.rb (hunk 2 above)
Comment (joshcooper) [reply lib/puppet/util.rb]: Since `path` is passed in, I don't want to mutate the callers variable.

File: lib/puppet/provider/augeas/augeas.rb (hunk 5 above)
Comment (joshcooper) [reply lib/puppet/provider/augeas/augeas.rb]: I didn't feel like changing it because the provider is removed from puppet in 6.0 and I didn't want to break anything.

File: lib/puppet/face/module/uninstall.rb [hunk 6]
```diff
@@ -61,7 +61,7 @@
     end
 
     when_invoked do |name, options|
-      name = name.gsub('/', '-')
+      name = name.tr('/', '-')
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply lib/puppet/face/module/uninstall.rb]: I'm reluctant to use mutating `tr` because of issues like PUP-7141. Since `name` is passed to us, I don't think we should mutate it.

File: lib/puppet/indirector/hiera.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (joshcooper) [reply lib/puppet/indirector/hiera.rb]: I also prefer the style you're talking about, but that kind of refactor is more than I wanted to do. If we did do something like that, then I'd only want to do it in `master` to avoid massive merge conflicts.

File: lib/puppet/type.rb (hunk 1 above)
Comment (joshcooper) [reply lib/puppet/type.rb]: There's a https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/BlockDelimiters cop for that, but it's out of scope for this PR.

File: lib/puppet/face/module/upgrade.rb (hunk 3 above)
Comment (joshcooper) [reply lib/puppet/face/module/upgrade.rb]: Same comment as https://github.com/puppetlabs/puppet/pull/7619#discussion_r306078172

File: lib/puppet/provider/package/rpm.rb [hunk 7]
```diff
@@ -335,8 +335,8 @@ def rpm_parse_evr(s)
       v = s[0,ri]
       r = s[ri+1,s.length]
       if arch = r.scan(ARCH_REGEX)[0]
-        a = arch.gsub(/\./, '')
-	r.gsub!(ARCH_REGEX, '')
+        a = arch.delete('.')
+	      r.gsub!(ARCH_REGEX, '')
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply lib/puppet/provider/package/rpm.rb]: Thanks! there was a tab in there. Updated

**Comment by joshcooper** [comment]
Need to resolve conflicts

File: spec/unit/type/file_spec.rb [hunk 8]
```diff
@@ -485,17 +488,19 @@
     describe "and a source is set" do
       it "should pass the already-discovered resources to recurse_remote" do
         file[:source] = File.expand_path(__FILE__)
-        allow(file).to receive(:recurse_local).and_return(:foo => "bar")
-        expect(file).to receive(:recurse_remote).with(:foo => "bar").and_return([])
+        allow(child).to receive(:[]).with(:path).and_return(name)
```
Comment (kris-bosland) [inline spec/unit/type/file_spec.rb]: Are these related to a RuboCop issue?

File: spec/unit/util/log/destinations_spec.rb [hunk 9]
```diff
@@ -46,13 +46,15 @@
...
       describe "with a normal file" do
+        let (:parent) { Pathname.new('/tmp') }
         let (:abspath) { '/tmp/log' }
         let (:relpath) { 'log' }
 
         it_behaves_like "file destination"
 
         it "logs an error if it can't chown the file owner & group" do
-          expect(File).to receive(:exists?).with(abspath).and_return(false)
+          allow(File).to receive(:exist?).with(parent).and_return(true)
```
Comment (kris-bosland) [inline spec/unit/util/log/destinations_spec.rb]: Was this test giving false positives before?

File: lib/puppet/pops/loader/null_loader.rb [hunk 10]
```diff
@@ -39,14 +39,12 @@ def loaded_entry(typed_name, check_dependencies = false)
...
 
-  # Has no entries on its own - always nil
   def get_entry(typed_name)
-    nil
-  end
-
-  # Finds nothing, there are no entries
-  def find(name)
-    nil
+    if @parent.nil?
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply lib/puppet/pops/loader/null_loader.rb]: I deleted it

File: spec/unit/util/log/destinations_spec.rb (hunk 9 above)
Comment (joshcooper) [reply spec/unit/util/log/destinations_spec.rb]: Ah earlier in the same method, we're using `Puppet::FileSystem.exist?` to check the parent directory, so I changed this one to match. I updated the commit message.

File: spec/unit/type/file_spec.rb (hunk 8 above)
Comment (joshcooper) [reply spec/unit/type/file_spec.rb]: I updated the commit summary to describe the test change.

File: spec/unit/indirector/resource/ral_spec.rb [hunk 11]
```diff
@@ -47,15 +47,15 @@
     end
 
     it "should convert ral resources into regular resources" do
-      my_resource = double("my user resource")
+      my_resource = double("my user resource", :title => "my user resource")
```
[the remark this answers is not in the document]
Comment (joshcooper) [reply spec/unit/indirector/resource/ral_spec.rb]: I updated the commit summary to describe the test change.
